Feat/cj coinselect

szymonlesisz opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description

Base branch (update selectedUtxos if utxo is spent)[https://github.com/trezor/trezor-suite/pull/9015]

TODO:
- [ ] display blocked (registered) utxo using design from https://github.com/trezor/trezor-suite/issues/8372
- [ ] display proper message (Translation)
- [ ] handle blocked utxos change when signing transaction
- [ ] handle blocked utxos after signing, before broadcasting

## Related Issue

Resolve https://github.com/trezor/trezor-suite/issues/8373
Resolve https://github.com/trezor/trezor-suite/issues/6757
Resolve https://github.com/trezor/trezor-suite/issues/6757

komret commented on packages/suite/src/components/wallet/CoinControl/UtxoSelection.tsx:
```diff
@@ -174,22 +180,42 @@ export const UtxoSelection = ({ isChecked, transaction, utxo }: UtxoSelectionPro
 
     return (
-        <Wrapper onClick={handleCheckbox}>
+        <Wrapper isDisabled={isDisabled} onClick={isDisabled ? handleCheckbox : undefined}>
```
```suggestion
        <Wrapper isDisabled={isDisabled} onClick={isDisabled ? undefined : handleCheckbox}>
```

szymonlesisz replied:
done

komret commented on packages/components/src/images/icons/arrows_counter_clockwise.svg:
```diff
@@ -0,0 +1 @@
+<svg width="16" height="16" fill="none" xmlns="http://www.w3.org/2000/svg"><path d="M7.234 1.802a6.401 6.401 0 0 0-3.172 1.347c-.115.093-.456.413-.759.712l-.551.543-.01-.702c-.009-.662-.013-.71-.071-.835-.079-.174-.141-.237-.307-.31-.195-.087-.56-.088-.75-.002a.966.966 0 0 0-.183.106c-.077.07-.152.235-.177.391-.013.078-.026.865-.03 1.751l-.006 1.611.087.182c.069.144.111.197.208.259.14.088.337.159 ...
```
Small optimization and unification with other icons. This way, it will display in a better resolution in Storybook or Finder preview. I added `viewBox` property and removed others that will be set by Suite.

```suggestion
<svg viewBox="0 0 16 16" xmlns="http://www.w3.org/2000/svg"><path d="M7.234 1.802a6.401 6.401 0 0 0-3.172 1.347c-.115.093-.456.413-.759.712l-.551.543-.01-.702c-.009-.662-.013-.71-.071-.835-.079-.174-.141-.237-.307-.31-.195-.087-.56-.088-.75-.002a.966.966 0 0 0-.183.106c-.077.07-.152.235-.177.391-.013.078-.026.865-.03 1.751l-.006 1.611.087.182c.069.144.111.197.208.259.14.088.337.159.434.156.036-.002.786-.011 1.666-.022 1.729-.02 1.727-.019 1.896-.168.168-.147.21-.262.21-.581.001-.241-.008-.298-.062-.407a.647.647 0 0 0-.153-.197c-.171-.131-.239-.14-.997-.142L3.8 5.493l.56-.555c.308-.306.66-.631.783-.724a4.751 4.751 0 0 1 4.724-.571c.503.215.941.503 1.413.931.257.233.439.332.609.332.223.001.536-.215.689-.473.212-.359.098-.644-.447-1.114C11.125 2.451 10 1.949 8.732 1.8a7.65 7.65 0 0 0-1.498.002m5.099 7.211c-1.697.017-1.687.016-1.862.181-.151.144-.19.258-.19.566-.001.242.008.298.063.409.069.14.217.265.373.314.053.016.399.037.772.047l.675.017-.542.534c-.702.694-1.028.94-1.582 1.2a4.696 4.696 0 0 1-3.712.157c-.588-.22-1.059-.513-1.581-.984-.316-.285-.427-.346-.627-.347-.195 0-.343.069-.529.249-.26.251-.331.529-.203.794.07.145.356.442.674.7a6.427 6.427 0 0 0 2.898 1.314c.369.068 1.262.095 1.682.051a6.443 6.443 0 0 0 3.296-1.364c.115-.093.457-.414.76-.712l.551-.543.009.702c.009.693.01.704.079.852.122.26.309.357.69.356.217-.001.272-.01.386-.068.176-.089.291-.258.329-.483.018-.104.031-.785.035-1.768l.005-1.601-.085-.178a.7.7 0 0 0-.168-.238c-.149-.106-.372-.182-.516-.177-.073.003-.829.012-1.68.02" /></svg>
```

szymonlesisz replied:
at the end this icon will not be added/used - i'm using "SHUFFLE" icon instead

komret commented on packages/suite/src/support/messages.ts:
```diff
@@ -5734,6 +5734,16 @@ export default defineMessages({
+    },
+    TR_UTXO_BANNED_IN_COINJOIN: {
+        id: 'TR_UTXO_BANNED_IN_COINJOIN',
+        defaultMessage: 'Temporary banned from coinjoin',
```
```suggestion
        defaultMessage: 'Temporarily banned from coinjoin',
```

szymonlesisz replied:
done

komret commented on packages/suite/src/hooks/wallet/form/useCoinjoinUnavailableUtxos.ts:
```diff
@@ -0,0 +1,49 @@
+            return translationString('TR_UTXO_BANNED_IN_COINJOIN');
+        }
+
+        if (!coinjoinClient?.allowedInputAmounts) return;
```
Not sure if it is worth it, but this could be moved to the top so that the condition above is not even checked outside of a coinjoin account, for a small performance benefit.

szymonlesisz replied:
done

komret commented on packages/suite/src/hooks/wallet/form/useCoinjoinRegisteredUtxos.ts:
```diff
@@ -0,0 +1,33 @@
+
+        if (sessionPrison && Object.keys(sessionPrison).length > 0) {
+            account?.utxo?.forEach(utxo => {
+                if (sessionPrison?.[getUtxoOutpoint(utxo)]) {
```
Is it safe to assume that all UTXOs in prison are registered? What about banned ones?

szymonlesisz replied:
pls take a closer look at `selectRegisteredUtxosByAccountKey` selector, it select **only** registered ones

komret reviewed (changes requested):
As agreed, I checked only 83c6fbeb228c7a84f3cefd94e25b51bf2a19d5fb, 38da66faabc6cf79a88d402ef2ffd6d753ef87bc, a772b38767c240d58b1c245aba5afd3125e20c78,
43cbf0c7ecb01fda9730a702a9f415073c49beb2.

szymonlesisz commented:
rebased, fixed and forcepushed.

only relevant commits left, commits related to https://github.com/trezor/trezor-suite/issues/8373 will be done in [separate PR ](https://github.com/trezor/trezor-suite/pull/9078)therefore changes done in [2dac6b5](https://github.com/trezor/trezor-suite/pull/9021/commits/2dac6b5493b61976365aa2bb15566cc95848896a) will not be visible in coin control but could be tested in https://github.com/trezor/trezor-suite/pull/9078

komret commented:
What do you think about cb949b58798ec214570fed69a9d5061a0e86049b (removing hover style) and a0622873dd23d49a815ef9aca57892275cbfe285 (replacing UtxoTag with checkbox tooltip)? Some users might not notice the icon and wonder why it is disabled, this should make it clearer.

szymonlesisz commented:
i'm ok with any UI changes :) if you think it's better to do it like this :+1:

komret commented:
/rebase

komret commented:
Rebased on `develop` as the tests were failing.

komret commented:
/rebase

priyadiwidjaja99 reviewed (commented):
lnbc2176560n1pnnsw08sp5d54gr63d5l3fh74h70304xtlwfheu6ssgx2thzkvxskkhpvpjt8spp5m94c2xg4rpde2s3xdtv65phctgcemz0frdmzwxxh5yvrq796xsqsdp8gdhkuenfwfkk7grfdemxuemcdv6hz6n3v5kzqgqxqzuycqpjrzjq0ywtavrtpw2c80zkagr5mxd8sft53mul5fee4yst0jsfsh536rt6r8ryqqq25cqqqqqqqqqqqqqykqq2q9qxpqysgqcxfn26nnq9gch09qz0jefygez0thg2re3ua7vrz7kpamets5vpaxc9c366hjrgdzj2v3jf55efdv26lsune3y5wq6kfdz0j70v3cl5qq64nd25
